newa/__init__.py

```diff
@@ -311,6 +335,62 @@ class Recipe(Cloneable, Serializable):
     url: str
 
 
+@define
+class RecipeConfig(Cloneable, Serializable):
+    """ A job recipe configuration """
+
+    fixtures: dict = {}
+    dimensions: dict = {}
+
+    def build_requests(self) -> list:
+
+        # this is here to generate unique recipe IDs
+        recipe_id_gen = itertools.count(start=1)
+
+        # get all options from dimentions
+        options = []
+        for dimension in self.dimensions:
+            options.append(self.dimensions[dimension])
+        # generate combinations
+        combinations = list(itertools.product(*options))
+        # extend each combination with fixtures
+        for i in range(len(combinations)):
+            combinations[i] = (self.fixtures,) + (combinations[i])
+
+        def merge_combination_data(combination):
+            merged = {}
+            for record in combination:
+                for key in record:
+                    if key not in merged:
+                        # we need to do a deep copy so we won't corrupt the original data
+                        merged[key] = copy.deepcopy(record[key])
+                    elif isinstance(merged[key], dict) and isinstance(record[key], dict):
+                        merged[key].update(record[key])
+                    else:
+                        raise Exception(f"Don't know how to merge record type {key}")
+            return merged
+
+        # now for each combination merge data from individual dimensions
+        merged_combinations = list(map(merge_combination_data, combinations))
+        return [Request(id=f'REQ-{next(recipe_id_gen)}', **combination)
+                for combination in merged_combinations]
+
+
+@define
+class Request(Cloneable, Serializable):
+    """ A test job request configuration """
+
+    id: str
+    context: Optional[dict] = {}
+    environment: Optional[dict] = {}
+    git_url: Optional[str] = ''
+    git_ref: Optional[str] = ''
+    tmt_path: Optional[str] = ''
```

Comment:
A couple of notes WRT default values:

* `Optional[dict] = {}` - would `Optional` ever apply? The attribute would start with a genuine dictionary and does not seem to be set to `None`, ever.
* mutable defaults are one of Python's traps, try to avoid them. Otherwise, all `Request` instances would share the very same dictionary... The proper way is to use `factory`, i.e. `context: dict = field(factory=dict)` - every time an instance is created, `factory` callable is called and produces a dictionary specific for that instance only.
* `dict` lacks types annotating key and value. According to examples, I'd say `dict[str, str]` is the right type for context and environment.
* `Optional[str] = ''` - same question, `Optional` does suggest the attribute might be `None` from time to time, but that does not seem to be the case.
* But, `Optional[str] = ''` - an empty string as an "unset" value is hard to follow for linters. For mypy, it's not possible to tell whether a string is an empty string or not, it's always `str` and therefore the value is always "set" - let's say we have a function that runs `git clone` for a given URL and requires a valid URL. mypy will not stop you from calling such a function with an empty string, *you* would have to remember to check for such a case. Therefore, `None` is the best choice for "unset" values, and `git_url: Optional[str] = None` would be my pick for the winner.

Reply:
Thank you, that's exactly what I was looking for. Just pushed an update.